model/riscv_vmem.sail

```diff
@@ -0,0 +1,854 @@
+                      -> xlenbits = {       // new legal value of satp
+  if sizeof(xlen) == 32 then {
+    // The slice and extend ops below are no-ops when xlen==32,
+    // but appease the type-checker when xlen==64 (when this code is not executed!)
+    let o32      : bits(32) = o[31 .. 0];
+    let v32      : bits(32) = v[31 .. 0];
+    let new_satp : bits(32) = legalize_satp32(a, o32, v32);
+    zero_extend(new_satp);
+  } else if sizeof(xlen) == 64 then {
+    // The extend and truncate ops below are no-ops when xlen==64,
+    // but appease the type-checker when xlen==32 (when this code is not executed!)
+    let o64      : bits(64) = zero_extend(o);
+    let v64      : bits(64) = zero_extend(v);
+    let new_satp : bits(64) = legalize_satp64(a, o64, v64);
+    truncate(new_satp, sizeof(xlen))
+  } else
+    internal_error(__FILE__, __LINE__, "Unsupported xlen" ^ string_of_int(sizeof(xlen)))
+}
+
+// ----------------
+// Fields of SATP
+
+// ASID is 9b in Sv32, 16b in Sv39/Sv48/Sv57: we use 16b for both
+// PRIVATE
+function satp_to_asid(sv_params : SV_Params, satp_val : xlenbits) -> bits(16) = {
+  // This extend op is a no-op when xlen==64, extends when xlen==32
+  let satp_64b : bits(64) = zero_extend(satp_val);
+  let mask_64b : bits(64) = zero_extend(ones(sv_params.satp_asid_size_bits));
+  let asid_64b : bits(64) = (satp_64b >>  sv_params.satp_asid_lsb_index) & mask_64b;
+  asid_64b[15 .. 0]
+}
+
+// Result is 64b to cover both RV32 and RV64 addrs
+// PRIVATE
+function satp_to_PT_base(sv_params : SV_Params, satp_val : xlenbits) -> bits(64) = {
+  // This extend op is a no-op when xlen==64, extends when xlen==32
+  let satp_64b : bits(64) = zero_extend(satp_val);
+  let mask_64b : bits(64) = zero_extend(ones(sv_params.satp_ppn_size_bits));
+  let ppn_64b  : bits(64) = (satp_64b >>  sv_params.satp_ppn_lsb_index) & mask_64b;
+  ppn_64b << pagesize_bits
+}
+
+// Compute address translation mode from SATP register
+// PRIVATE
+function translationMode(priv : Privilege) -> SATPMode = {
+  if priv == Machine then
+    Sbare
+  else if sizeof(xlen) == 32 then
+    match satp[31] {
+      bitzero => Sbare,
+      bitone  => Sv32
+    }
+  else if sizeof(xlen) == 64 then {
+    // This extend op is a no-op when xlen==64,
+    // but appeases the type-checker when xlen==32 (when this code is not executed!)
+    let satp_64b : bits(64) = zero_extend(satp);
+    // Translation mode is based on mstatus.SXL, which could be RV32 when xlen==64
+    let arch = architecture(get_mstatus_SXL(mstatus));
+    match arch {
+      Some(RV64) => { let mbits : bits(4) = satp[63 .. 60];
+                      match satp64Mode_of_bits(RV64, mbits) {    // see riscv_types.sail
+                        Some(m) => m,
+                        None()  => internal_error(__FILE__, __LINE__,
+                                                  "invalid RV64 translation mode in satp")
+                      }
+                    },
+      Some(RV32) => match satp[31] {
```

Comment:
Ditto.

Reply:
Ok. Fixed (although note that here, in RV64 mode, satp is 64bits, so can’t directly be an arg to Mk_Satp32).